Src/Support/Google.Apis.Core/Http/HttpClientFromMessageHandlerFactory.cs

```diff
@@ -0,0 +1,198 @@
+            }
+            else if (configuredMessageHandler.PerformsAutomaticDecompression)
+            {
+                // Complex: we want to use a simple handler with no interception but with built-in decompression
+                // for requests that wouldn't perform interception anyway, and a longer chain for interception cases.
+
+                var noCompressionMessageHandler = DisallowDecompressionOptions.CheckItMatches(_httpMessageHandlerFactory(DisallowDecompressionOptions));
+
+                return new TwoWayDelegatingHandler(
+                    // Normal handler (with built-in decompression) when there's no interception.
+                    configuredMessageHandler.MessageHandler,
+                    // Alternative handler for requests that might be intercepted, and need that interception to happen
+                    // before decompression.
+                    new GzipDeflateHandler(new StreamInterceptionHandler(noCompressionMessageHandler.MessageHandler)),
+                    request => StreamInterceptionHandler.GetInterceptorProvider(request) != null);
+            }
+            else
+            {
+                // Simple: we have to create our own decompression handler anyway, so there's still just a single chain.
+                var interceptionHandler = new StreamInterceptionHandler(configuredMessageHandler.MessageHandler);
+                return new GzipDeflateHandler(interceptionHandler);
+            }
+        }
+
+        /// <summary>
+        /// Specifies the configuration options for a message handler.
+        /// </summary>
+        public class HttpMessageHandlerOptions
+        {
+            /// <summary>
+            /// Whether the message handler built from these options
+            /// may perform automatic decompression or not.
+            /// If set to true, the message handler may or may not perform automatic decompression.
+            /// If set to false, the message handler must not perform automatic decompression.
+            /// </summary>
+            public bool MayPerformDecompression { get; }
+
+            /// <summary>
+            /// Whether the message handler built from these options
+            /// may handle redirects or not. Redirects that are not handled
+            /// should buble up the handlers chain.
+            /// If set to true, the message handler may or may not handle redirects.
+            /// If set to false, the message handler must not handle redirects.
+            /// </summary>
+            public bool MayHandleRedirects { get; }
+
+            internal HttpMessageHandlerOptions(bool mayPerformDecompression, bool mayHandleRedirects)
+            {
+                MayPerformDecompression = mayPerformDecompression;
+                MayHandleRedirects = mayHandleRedirects;
+            }
+
+            internal ConfiguredHttpMessageHandler CheckItMatches(ConfiguredHttpMessageHandler handler)
```

Comment:
I'm not sure how useful this is, given that we're already basically trusting what the user has told us in ConfiguredHttpMessageHandler. It's not like we're examining the underlying handler, unless I'm missing something.

Reply:
As agreed offline, leaving this in as a way to check that the message handler we are given matches what we requested. Still there's a level of trust we have in user code, because we are not checking the actual handler, just the values here, which may or may not reflect what the handler actually does. This check will be useful for catching accidental missatches. For on-purpose missmatches, we simply can't guarantee that service calls will work as expected.